File: examples/typescript/multi_session.ts
/**
 * EdgeBox TypeScript Multi-Session Example
 *
 * Demonstrates running multiple tasks in the EdgeBox sandbox concurrently.
 *
 * Prerequisites:
 *   1. EdgeBox is running with Docker Desktop
 *   2. MCP server is healthy (check the EdgeBox dashboard)
 *   3. Install dependencies: npm install
 *
 * Usage:
 *   npx tsx multi_session.ts
 */

import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";
import { SSEClientTransport } from "@modelcontextprotocol/sdk/client/sse.js";

const EDGEBOX_MCP_URL = "http://localhost:8888/mcp";

async function createClient(): Promise<Client> {
  const client = new Client(
    { name: "edgebox-ts-multi-session", version: "1.0.0" },
    { capabilities: {} }
  );

  try {
    const transport = new StreamableHTTPClientTransport(
      new URL(EDGEBOX_MCP_URL)
    );
    await client.connect(transport);
  } catch {
    const sseTransport = new SSEClientTransport(
      new URL("http://localhost:8888/sse")
    );
    await client.connect(sseTransport);
  }

  return client;
}

async function runTask(taskName: string, code: string) {
  const client = await createClient();

  try {
    console.log(`[${taskName}] Executing code...`);
    const result = await client.callTool({
      name: "execute_python",
      arguments: { code },
    });
    console.log(`[${taskName}] Result:`, result.content);
    return result;
  } finally {
    await client.close();
  }
}

async function main() {
  console.log("=".repeat(60));
  console.log("EdgeBox Multi-Session Example");
  console.log("=".repeat(60));
  console.log("\nRunning two tasks concurrently...\n");

  // Task 1: Data processing
  const dataTask = `
import statistics
data = [23, 45, 67, 12, 89, 34, 56, 78, 90, 11]
print(f"Mean: {statistics.mean(data):.2f}")
print(f"Median: {statistics.median(data):.2f}")
print(f"Stdev: {statistics.stdev(data):.2f}")
`;

  // Task 2: System info
  const systemTask = `
import platform, os
print(f"OS: {platform.system()} {platform.release()}")
print(f"Architecture: {platform.machine()}")
print(f"Python: {platform.python_version()}")
print(f"CPU count: {os.cpu_count()}")
`;

  // Run tasks concurrently
  await Promise.all([
    runTask("data-analysis", dataTask),
    runTask("system-info", systemTask),
  ]);

  console.log("\nAll sessions completed!");
}

main().catch(console.error);

File: examples/python/multi_session.py
"""
EdgeBox Python Multi-Session Example

Demonstrates running multiple isolated sandbox sessions concurrently.
Each session has its own filesystem and execution environment.

Prerequisites:
  1. EdgeBox is running with Docker Desktop
  2. MCP server is healthy (check the EdgeBox dashboard)
  3. Install dependencies: pip install -r requirements.txt

Usage:
  python multi_session.py
"""

import asyncio
from fastmcp import Client


EDGEBOX_BASE_URL = "http://localhost:8888/mcp"


async def run_session(session_name: str, code: str):
    """Run code in an isolated EdgeBox session."""
    # Each session gets its own isolated sandbox via the x-session-id header.
    # Pass custom headers through the Client to route to different containers.
    url = f"{EDGEBOX_BASE_URL}"
    client = Client(url)

    async with client:
        print(f"[{session_name}] Executing code...")
        result = await client.call_tool("execute_python", {"code": code})
        print(f"[{session_name}] Result: {result}")
        return result


async def main():
    print("=" * 60)
    print("EdgeBox Multi-Session Example")
    print("=" * 60)
    print()
    print("Running two tasks in the same EdgeBox sandbox...\n")

    # Task 1: Data processing
    data_task = """
import statistics
data = [23, 45, 67, 12, 89, 34, 56, 78, 90, 11]
print(f"Mean: {statistics.mean(data):.2f}")
print(f"Median: {statistics.median(data):.2f}")
print(f"Stdev: {statistics.stdev(data):.2f}")
"""

    # Task 2: System info
    system_task = """
import platform, os
print(f"OS: {platform.system()} {platform.release()}")
print(f"Architecture: {platform.machine()}")
print(f"Python: {platform.python_version()}")
print(f"CPU count: {os.cpu_count()}")
"""

    # Run tasks concurrently
    await asyncio.gather(
        run_session("data-analysis", data_task),
        run_session("system-info", system_task),
    )

    print("\nAll sessions completed!")


if __name__ == "__main__":
    asyncio.run(main())
